### DISCUSSION TITLE
Automatically generate slug based on other field

### DISCUSSION THREAD

**Discussion by ndaidong** [directus/directus#15195] [Feature Requests]
A unique slug is better for SEO than id/uuid. Should we borrow this feature from Contentful?

![Screenshot from 2022-08-21 10-19-20](https://user-images.githubusercontent.com/1809619/185773957-d33a9c5f-e5a2-4883-968e-bf7c5ba8de05.png)

**Comment by rijkvanzanten** [comment]
Also see https://github.com/directus/directus/discussions/3079, https://github.com/directus/directus/discussions/4231 🙂

**Comment by azrikahar** [comment]
In the mean time, the addition of Run Script operation in Flows should help out in this use case, even if it's not as direct as a dedicated slug field/interface.

Here's an example:

1. Make sure to install `@sindresorhus/slugify` version `1.1.0` (because v2 moved to pure ESM)
2. Configure the env variable `FLOWS_EXEC_ALLOWED_MODULES` to opt in the modules that are allowed to be used for security reasons. In this case, I've configured it as `FLOWS_EXEC_ALLOWED_MODULES=array:@sindresorhus/slugify`.
3. Create an `articles` collection, with `title` string field, and a `slug` readonly string field.
4. Create a Flow with Event Hook type Filter trigger on create and/or update of `articles` collection.
5. Add a Run Script operation with the following code:

    ```js
    const slugify = require('@sindresorhus/slugify');
    
    module.exports = async function(data) {
	    const slug = slugify(data.$trigger.payload.title);
	    return {...data.$trigger.payload, slug};
    }
    ```
    
    ⚠ Make sure there is the "The following Node modules can be used" message below the Code input to indicate you have configured the env variable `FLOWS_EXEC_ALLOWED_MODULES` successfully

6. Create an `articles` item by filling up the `title` field, then the `slug` field should auto populate with the slugified value of `title` field.

Quick video to show the example above:

https://user-images.githubusercontent.com/42867097/189595070-1be0868c-0d23-4e41-97f4-d90c3015d091.mp4

**Reply by leosin** [reply]
I did it every step of the way

create post > add title > publish.  
that auto generate slug ok.

when edit post , with update new title , it works

but without update title ,it tip error


![image](https://user-images.githubusercontent.com/3676670/191688353-4e14ad02-4d8a-4df1-96a5-d9bdda5b5395.png)

**Reply by azrikahar** [reply]
@leosin I believe the error is from `slugify()` as it does receive undefined value when you're not updating title. You should be able to resolve this with an additional guard statement such as `if (!data.$trigger.payload.title) return data.$trigger.payload;` (note this will also be true for empty title for example) or `if (data.$trigger.payload.title === undefined) return data.$trigger.payload;`.

Example usage:

```js
const slugify = require('@sindresorhus/slugify');

module.exports = async function(data) {
    if (data.$trigger.payload.title === undefined) return data.$trigger.payload;
    const slug = slugify(data.$trigger.payload.title);
    return {...data.$trigger.payload, slug};
}
```

With that said, the original code was meant as a quick proof of concept to show how it works, so it may not cover other scenarios! 😄 Do modify it accordingly as you see fit based on your use case.

**Reply by leosin** [reply]
😄 it works ,I learned a lot from this example, thank you very much . ❤️

**Comment by rijkvanzanten** [comment]
Closing this as duplicate of https://github.com/directus/directus/discussions/4231

**Comment by s-kennedy** [comment]
If anyone is still looking for a way to do this, this is what worked for me in version 10.8.2:

1. Create a new flow and select your event trigger (I'm using `items.create`) 
2. Select the "Filter (blocking)" type
3. Select "Data of last operation" for the response body

Add the "run script" operation and paste in the following (which is lightly adapted from this [directus guide](https://docs.directus.io/guides/flows/slugify-text-with-run-script.html)):

```
module.exports = async function (data) {
	// Index data to get the string you want to slugify
	// Assign it to the "text" var below.
	const text = data.$trigger.payload.name;

	const slug = text
		.toLowerCase()
		.trim()
		.replace(/[^\w\s-]/g, '')
		.replace(/[\s_-]+/g, '-')
		.replace(/^-+|-+$/g, '');

	return { 
        ...data.$trigger.payload, slug
    }
};
```
It took me wayyy to long to figure this out so I hope it saves someone some time.

**Comment by vwasteels** [comment]
Hello there,

These workaround are fine, but how do you manage uniqueness of the rendered slug ?

Thanks
